File: 200-number-of-islands/200-number-of-islands.cpp
class Solution {
public:
    bool flags[301][301];
    int numIslands(vector<vector<char>>& grid) {
        int rows = grid.size();
        int columns = grid[0].size();
        int count = 0;
        
        for(int i=0;i<rows;i++){
            for(int j=0;j<columns;j++){
                if(grid[i][j] == '1' && (flags[i][j] == false))
                    count += visit_island(i, j, grid, rows, columns);
            }
        }
        
        return count;
    }
    
    int visit_island(int i, int j, vector<vector<char>>& grid, int rows, int columns){
        pair<int,int> front;
        queue<pair<int,int>>q;
        q.push(make_pair(i,j));
        while(!q.empty()){
            front = q.front();
            q.pop();
            
            i = front.first;
            j = front.second;
            
            if(i <0 || j <0 || i >= rows || j >= columns || flags[i][j]) continue;
            if(grid[i][j] == '0') continue;
            
            flags[i][j] = true;
            
            q.push(make_pair(i-1, j));
            q.push(make_pair(i+1, j));
            q.push(make_pair(i, j-1));
            q.push(make_pair(i, j+1));
        }
        return 1;
    }
};

File: 200-number-of-islands/200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m=len(grid)
        n=len(grid[0])
        n_ilands=0
        
        def change(i, j):
            grid[i][j]="-1"
                
                
            if i>0 and grid[i-1][j]=="1": change(i-1,j)
            if j>0 and grid[i][j-1]=="1": change(i,j-1)
            if i<m-1 and grid[i+1][j]=="1": change(i+1,j)
            if j<n-1 and grid[i][j+1]=="1": change(i,j+1)
        
        for i in range(m):
            for j in range(n):
                if grid[i][j]=="1":
                    n_ilands+=1
                    change(i,j)
        
        return n_ilands
                
        